library/src/main/java/com/getstream/sdk/chat/adapter/MessageViewHolderFactory.kt

```diff
@@ -0,0 +1,103 @@
+package com.getstream.sdk.chat.adapter
+
+import android.view.ViewGroup
+import com.getstream.sdk.chat.R
+import com.getstream.sdk.chat.adapter.MessageListItem.DateSeparatorItem
+import com.getstream.sdk.chat.adapter.MessageListItem.MessageItem
+import com.getstream.sdk.chat.adapter.MessageListItem.ThreadSeparatorItem
+import com.getstream.sdk.chat.adapter.MessageListItem.TypingItem
+import com.getstream.sdk.chat.model.ModelType
+import io.getstream.chat.android.client.models.Attachment
+import io.getstream.chat.android.client.models.Message
+
+/**
+ * Allows you to easily customize message rendering or message attachment rendering
+ */
+open class MessageViewHolderFactory {
+    companion object {
+        const val MESSAGEITEM_DATE_SEPARATOR = 1
+        const val MESSAGEITEM_MESSAGE = 2
+        const val MESSAGEITEM_TYPING = 3
+        const val MESSAGEITEM_THREAD_SEPARATOR = 4
+        const val MESSAGEITEM_NOT_FOUND = 5
+
+        const val GENERIC_ATTACHMENT = 1
+        const val IMAGE_ATTACHMENT = 2
+        const val VIDEO_ATTACHMENT = 3
+        const val FILE_ATTACHMENT = 4
+    }
+
+    open fun getMessageViewType(messageListItem: MessageListItem?): Int {
+        return when (messageListItem) {
+            is DateSeparatorItem -> MESSAGEITEM_DATE_SEPARATOR
+            is TypingItem -> MESSAGEITEM_TYPING
+            is MessageItem -> MESSAGEITEM_MESSAGE
+            is ThreadSeparatorItem -> MESSAGEITEM_THREAD_SEPARATOR
+            else -> MESSAGEITEM_NOT_FOUND
+        }
+    }
+
+    open fun getAttachmentViewType(
+        message: Message,
+        mine: Boolean,
+        position: Position,
+        attachments: List<Attachment>,
+        attachment: Attachment
+    ): Int {
```

Comment:
It seems that we don't need all those arguments

Reply:
As far as I can tell, the purpose of this class is to allow for customization (by overriding these methods as needed in a subclass), so there could be implementations of it by users where these arguments are being used for something. Not sure which ones - if any - would be safe to remove.

Reply:
It is true. But all those info could be only into the "item" that will be rendered, and we can be sure that all info needed for this type will be into its "itemClass". I mean, we could have different types of items to be rendered, and every of them will have only the needed data to be shown on the ViewHolder.
BTW. I can see this factory method is not used into `MessageListItemAdapter`, it is into `AttachmentListItemAdapter`, Maybe we can create another task to cover it instead of add lot of changes on this PR

Reply:
How about only keeping the `attachment` parameter then? Feels like that should always be enough to figure out the type, and then everything else can be taken into account later, when binding date into the ViewHolder.

Reply:
Sound good, but maybe we can wrap it into some class like `AttachmentItem` that will live into our `AttachmentListItemAdater`, it would allow us to add extra info to be rendered on the `ViewHolder` without break our public API

Reply:
For now, I've removed the extra parameters, and kept just the `Attachment`. Will follow with a wrapper class around `Attachment` in a separate PR.